packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/EnumTypeProvider.cs

```diff
@@ -0,0 +1,386 @@
+            }
+            IsAccessibilityOverridden = input.Accessibility != null;
+
+            var isExtensible = input.IsExtensible;
+            if (ExistingType != null)
+            {
+                isExtensible = ExistingType.TypeKind switch
+                {
+                    TypeKind.Enum => false,
+                    TypeKind.Struct => true,
+                    _ => throw new InvalidOperationException(
+                        $"{ExistingType.ToDisplayString()} cannot be mapped to enum," +
+                        $" expected enum or struct got {ExistingType.TypeKind}")
+                };
+
+                _typeMapping = sourceInputModel?.CreateForModel(ExistingType);
+            }
+
+            if (isExtensible)
+            {
+                // extensible enums are implemented by readonly structs
+                DeclarationModifiers |= TypeSignatureModifiers.Partial | TypeSignatureModifiers.ReadOnly;
+            }
+
+            Name = input.Name.ToCleanName();
+            Namespace = GetDefaultModelNamespace(CodeModelPlugin.Instance.Configuration.Namespace);
+            IsExtensible = isExtensible;
+            ValueType = CodeModelPlugin.Instance.TypeFactory.CreateCSharpType(input.EnumValueType);
+            IsStringValueType = ValueType.Equals(typeof(string));
+            IsIntValueType = ValueType.Equals(typeof(int)) || ValueType.Equals(typeof(long));
+            IsFloatValueType = ValueType.Equals(typeof(float)) || ValueType.Equals(typeof(double));
+            IsNumericValueType = IsIntValueType || IsFloatValueType;
+
+            Description = input.Description;
+
+            _valueField = new FieldDeclaration(FieldModifiers.Private | FieldModifiers.ReadOnly, ValueType, "_value");
+            SerializationMethodName = IsStringValueType && IsExtensible ? nameof(object.ToString) : $"ToSerial{ValueType.Name.FirstCharToUpperCase()}";
+            DeserializationMethodName = $"To{Name}";
+
+            if (!IsExtensible)
+            {
+                Serialization = new EnumTypeSerializationProvider(this, sourceInputModel);
+            }
+        }
+
+        private readonly FieldDeclaration _valueField;
+
+        public CSharpType ValueType { get; }
+        public bool IsExtensible { get; }
+        public bool IsIntValueType { get; }
+        public bool IsFloatValueType { get; }
+        public bool IsStringValueType { get; }
+        public bool IsNumericValueType { get; }
+        public string? Description { get; }
+        public override string Name { get; }
+        public override string Namespace { get; }
+        protected override TypeKind TypeKind => IsExtensible ? TypeKind.Struct : TypeKind.Enum;
+        public bool IsAccessibilityOverridden { get; }
+        internal string SerializationMethodName { get; }
+        internal string DeserializationMethodName { get; }
+
+        public EnumTypeSerializationProvider? Serialization { get; }
```

Comment:
Lets make this return type TypeProvider so we can make EnumTypeSerializationProvider internal.